File: csharp/uwapi-cs/map.cs
using System;
using System.Collections.Generic;
using System.Runtime.InteropServices;

namespace Unnatural
{
    using OverviewFlags = Interop.UwOverviewFlags;

    public struct Vector3
    {
        public float x, y, z;
    }

    public static class Map
    {
        public static string Name()
        {
            return name;
        }

        public static string Guid()
        {
            return guid;
        }

        public static IReadOnlyList<Vector3> Positions()
        {
            return positions;
        }

        public static IReadOnlyList<Vector3> Ups()
        {
            return ups;
        }

        public static IReadOnlyList<IReadOnlyList<uint>> Neighbors()
        {
            return neighbors;
        }

        public static IReadOnlyList<uint> Neighbors(uint position)
        {
            return neighbors[(int)position];
        }

        public static IReadOnlyList<byte> Terrains()
        {
            return terrains;
        }

        public static IReadOnlyList<OverviewFlags> Overview()
        {
            return overview;
        }

        public static uint[] Entities(uint position)
        {
            Interop.UwIds ns = new Interop.UwIds();
            Interop.uwOverviewIds(position, ref ns);
            return InteropHelpers.Ids(ns);
        }

        public static uint[] AreaRange(Vector3 point, float radius)
        {
            Interop.UwIds tiles = new Interop.UwIds();
            Interop.uwAreaRange(point.x, point.y, point.z, radius, ref tiles);
            return InteropHelpers.Ids(tiles);
        }

        public static uint[] AreaConnected(uint position, float radius)
        {
            Interop.UwIds tiles = new Interop.UwIds();
            Interop.uwAreaConnected(position, radius, ref tiles);
            return InteropHelpers.Ids(tiles);
        }

        public static uint[] AreaNeighborhood(uint position, float radius)
        {
            Interop.UwIds tiles = new Interop.UwIds();
            Interop.uwAreaNeighborhood(position, radius, ref tiles);
            return InteropHelpers.Ids(tiles);
        }

        public static uint[] AreaExtended(uint position, float radius)
        {
            Interop.UwIds tiles = new Interop.UwIds();
            Interop.uwAreaExtended(position, radius, ref tiles);
            return InteropHelpers.Ids(tiles);
        }

        public static bool TestVisible(Vector3 a, Vector3 b)
        {
            return Interop.uwTestVisible(a.x, a.y, a.z, b.x, b.y, b.z);
        }

        public static bool TestShooting(uint shooterPosition, uint shooterProto, uint targetPosition, uint targetProto)
        {
            return Interop.uwTestShooting(shooterPosition, shooterProto, targetPosition, targetProto);
        }

        public static float DistanceLine(uint ai, uint bi)
        {
            Vector3 a = positions[(int)ai];
            Vector3 b = positions[(int)bi];
            float x = a.x - b.x;
            float y = a.y - b.y;
            float z = a.z - b.z;
            x *= x;
            y *= y;
            z *= z;
            return (float)Math.Sqrt(x + y + z);
        }

        public static float DistanceEstimate(uint a, uint b)
        {
            return Interop.uwDistanceEstimate(a, b);
        }

        public static float Yaw(uint a, uint b)
        {
            return Interop.uwYaw(a, b);
        }

        public static bool TestConstructionPlacement(uint constructionPrototype, uint position)
        {
            return Interop.uwTestConstructionPlacement(constructionPrototype, position);
        }

        public static uint FindConstructionPlacement(uint constructionPrototype, uint position)
        {
            return Interop.uwFindConstructionPlacement(constructionPrototype, position);
        }

        static string name;
        static string guid;
        static readonly List<Vector3> positions = new List<Vector3>();
        static readonly List<Vector3> ups = new List<Vector3>();
        static readonly List<uint[]> neighbors = new List<uint[]>();
        static readonly List<byte> terrains = new List<byte>();
        static OverviewFlags[] overview = new OverviewFlags[0];

        static void Load()
        {
            Console.WriteLine("loading map");

            positions.Clear();
            ups.Clear();
            neighbors.Clear();
            terrains.Clear();

            {
                Interop.UwMapInfo info = new Interop.UwMapInfo();
                Interop.uwMapInfo(ref info);
                name = Marshal.PtrToStringAnsi(info.name);
                guid = Marshal.PtrToStringAnsi(info.guid);
                Console.WriteLine("map name: " + name);
                Console.WriteLine("map guid: " + guid);
            }

            uint count = Interop.uwTilesCount();
            Interop.UwTile tile = new Interop.UwTile();
            for (uint i = 0; i < count; ++i)
            {
                Interop.uwTile(i, ref tile);
                {
                    Vector3 p = new Vector3();
                    p.x = tile.position[0];
                    p.y = tile.position[1];
                    p.z = tile.position[2];
                    positions.Add(p);
                }
                {
                    Vector3 u = new Vector3();
                    u.x = tile.up[0];
                    u.y = tile.up[1];
                    u.z = tile.up[2];
                    ups.Add(u);
                }
                {
                    uint[] tmp = new uint[tile.neighborsCount];
                    if (tile.neighborsCount > 0)
                        Marshal.Copy(tile.neighborsIndices, (int[])(object)tmp, 0, (int)tile.neighborsCount);
                    neighbors.Add(tmp);
                }
                terrains.Add(tile.terrain);
            }

            Console.WriteLine("map loaded");
        }

        static void Preparing(object sender, EventArgs e)
        {
            Load();
        }

        static void Updating(object sender, EventArgs e)
        {
            if (Interop.uwGameState() == 3)
            {
                Interop.UwOverviewExtract ex = new Interop.UwOverviewExtract();
                Interop.uwOverviewExtract(ref ex);
                if (overview.Length != ex.count)
                    overview = new OverviewFlags[ex.count];
                if (ex.count > 0)
                    Marshal.Copy(ex.flags, (int[])(object)overview, 0, (int)ex.count);
            }
            else
                overview = new OverviewFlags[0];
        }

        static Map()
        {
            Game.Preparing += Preparing;
            Game.Updating += Updating;
        }
    }

    public static class InteropHelpers
    {
        public static uint[] Ids(Interop.UwIds ids)
        {
            uint[] tmp = new uint[ids.count];
            if (ids.count > 0)
                Marshal.Copy(ids.ids, (int[])(object)tmp, 0, (int)ids.count);
            return tmp;
        }
    }
}

File: python/uw/map.py
import math

from .helpers import MapState
from .helpers import OverviewFlags


class Vector3:
    def __init__(self, x: float, y: float, z: float):
        self.x = x
        self.y = y
        self.z = z


class Map:
    def __init__(self, api, ffi, game):
        self._api = api
        self._ffi = ffi
        self._game = game

        self._game.add_map_state_callback(self._map_state_changed)
        self._game.add_update_callback(self._updating)

        self._name: str = ""
        self._guid: str = ""
        self._path: str = ""
        self._max_players: int = 0
        self._positions: list[Vector3] = []
        self._ups: list[Vector3] = []
        self._neighbors: list[list[int]] = []
        self._terrains: list[bytes] = []
        self._overview: list[OverviewFlags] = []

    def name(self) -> str:
        return self._name

    def guid(self) -> str:
        return self._guid

    def path(self) -> str:
        return self._path

    def max_players(self) -> int:
        return self._max_players

    def positions(self) -> list[Vector3]:
        return self._positions

    def ups(self) -> list[Vector3]:
        return self._ups

    def neighbors(self) -> list[list[int]]:
        return self._neighbors

    def neighbors_on_pos(self, pos: int) -> list[int]:
        return self._neighbors[pos]

    def terrains(self) -> list[bytes]:
        return self._terrains

    def overview(self) -> list[OverviewFlags]:
        return self._overview

    def entities(self, position: int) -> list[int]:
        ns = self._ffi.new("struct UwIds *")
        self._api.uwOverviewIds(position, ns)
        return self._ffi.unpack(ns)

    def area_range(self, point: Vector3, radius: float) -> list[int]:
        tiles = self._ffi.new("struct UwIds *")
        self._api.uwAreaRange(point.x, point.y, point.z, radius, tiles)
        return self._ffi.unpack(tiles)

    def area_connected(self, position: int, radius: float) -> list[int]:
        tiles = self._ffi.new("struct UwIds *")
        self._api.uwAreaConnected(position, radius, tiles)
        return self._ffi.unpack(tiles)

    def area_neighborhood(self, position: int, radius: float) -> list[int]:
        tiles = self._ffi.new("struct UwIds *")
        self._api.uwAreaNeighborhood(position, radius, tiles)
        return self._ffi.unpack(tiles)

    def area_extended(self, position: int, radius: float) -> list[int]:
        tiles = self._ffi.new("struct UwIds *")
        self._api.uwAreaExtended(position, radius, tiles)
        return self._ffi.unpack(tiles)

    def test_visible(self, a: Vector3, b: Vector3) -> bool:
        return self._ffi.uwTestVisible(a.x, a.y, a.z, b.x, b.y, b.z)

    def test_shooting(self, shooter_position: int, shooter_proto: int, target_position: int, target_proto: int):
        return self._ffi.uwTestShooting(shooter_position, shooter_proto, target_position, target_proto)

    def distance_line(self, ai: int, bi: int) -> float:
        a: Vector3 = self._positions[ai]
        b: Vector3 = self._positions[bi]
        x: float = a.x - b.x
        y: float = a.y - b.y
        z: float = a.z - b.z
        x *= x
        y *= y
        z *= z
        return math.sqrt(x + y + z)

    def distance_estimate(self, a: int, b: int) -> float:
        return self._api.uwDistanceEstimate(a, b)

    def yaw(self, a: int, b: int) -> float:
        return self._api.uwDistanceEstimate(a, b)

    def test_construction_placement(self, construction_prototype: int, position: int) -> bool:
        return self._api.uwTestConstructionPlacement(construction_prototype, position)

    def find_construction_placement(self, construction_prototype: int, position: int) -> bool:
        return self._api.uwFindConstructionPlacement(construction_prototype, position)

    def _load(self):
        print("loading map")
        self._positions = []
        self._ups = []
        self._neighbors = []
        self._terrains = []
        self._overview = []

        info = self._ffi.new("struct UwMapInfo *")
        self._ffi.uwMapInfo(info)
        self._name = self._ffi.string(info.name)
        self._guid = self._ffi.string(info.guid)
        self._path = self._ffi.string(info.path)
        self._max_players = info.maxPlayers
        print(f"map name: {self._name}")
        print(f"map guid: {self._guid}")

        count = self._api.uwTilesCount()
        tile = self._ffi.new("struct UwTile *")
        for i in range(count):
            self._api.uwTile(i, tile)
            p = Vector3(tile.position[0], tile.position[1], tile.position[2])
            self._positions.append(p)
            u = Vector3(tile.up[0], tile.up[1], tile.up[2])
            self._ups.append(u)
            n = self._ffi.unpack(tile.neighborsIndices, tile.neighborsCount)
            if n:
                self._neighbors.append(n)
            self._terrains.append(tile.terrain)

        print("map loaded")

    def _map_state_changed(self, map_state: MapState):
        if map_state == MapState.Loaded:
            self._load()

    def _updating(self, stepping: bool):
        if stepping:
            ex = self._ffi.new("struct UwOverviewExtract *")
            self._api.uwOverviewExtract()
            if ex.count > 0:
                self._overview = [OverviewFlags(i) for i in self._ffi.unmarshal(ex.flags, ex.count)]
        else:
            self._overview = []
